lxd/project/permissions.go

```diff
@@ -1514,3 +1514,113 @@ func AllowSnapshotCreation(p *api.Project) error {
 }
+
+// GetRestrictedClusterGroups returns a slice of restricted cluster groups for the given project.
+func GetRestrictedClusterGroups(p *api.Project) []string {
+	return shared.SplitNTrimSpace(p.Config["restricted.cluster.groups"], ",", -1, true)
+}
+
+// AllowClusterMember returns nil if the given project is allowed to use the cluster member.
+func AllowClusterMember(p *api.Project, member *db.NodeInfo) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	if shared.IsTrue(p.Config["restricted"]) && len(clusterGroupsAllowed) > 0 {
+		for _, memberGroupName := range member.Groups {
+			if shared.StringInSlice(memberGroupName, clusterGroupsAllowed) {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("Project isn't allowed to use this cluster member: %q", member.Name)
+	}
+
+	return nil
+}
+
+// AllowClusterGroup returns nil if the given project is allowed to use the cluster group.
+func AllowClusterGroup(p *api.Project, groupName string) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	// Skip the check if the project is not restricted
+	if shared.IsFalseOrEmpty(p.Config["restricted"]) {
+		return nil
+	}
+
+	if len(clusterGroupsAllowed) > 0 && !shared.StringInSlice(groupName, clusterGroupsAllowed) {
+		return fmt.Errorf("Project isn't allowed to use this cluster group: %q", groupName)
+	}
+
+	return nil
+}
+
+// CheckTargetMember checks if the given member is allowed for the project.
+// If nil it also returns the resolved target member node information.
+func CheckTargetMember(p *api.Project, targetMember string, allMembers []db.NodeInfo) (*db.NodeInfo, error) {
+	// Find target member.
+	for _, potentialMember := range allMembers {
+		if potentialMember.Name == targetMember {
+			// If restricted groups are specified then check member is in at least one of them.
+			err := AllowClusterMember(p, &potentialMember)
+			if err != nil {
+				return nil, api.StatusErrorf(http.StatusForbidden, err.Error())
+			}
+
+			return &potentialMember, nil
+		}
+	}
+
+	return nil, api.StatusErrorf(http.StatusNotFound, "Cluster member %q not found", targetMember)
+}
+
+// CheckTargetGroup checks if the given group is allowed for the project.
+func CheckTargetGroup(ctx context.Context, tx *db.ClusterTx, p *api.Project, groupName string) error {
+	// If restricted groups are specified then check the requested group is in the list.
+	err := AllowClusterGroup(p, groupName)
+	if err != nil {
+		return api.StatusErrorf(http.StatusForbidden, err.Error())
+	}
+
+	// Check if the target group exists.
+	targetGroupExists, err := cluster.ClusterGroupExists(ctx, tx.Tx(), groupName)
+	if err != nil {
+		return err
+	}
+
+	if !targetGroupExists {
+		return api.StatusErrorf(http.StatusBadRequest, "Cluster group %q doesn't exist", groupName)
+	}
+
+	return nil
+}
+
+// CheckTarget checks if the given cluster target (member or group) is allowed.
+// If target is a cluster member, it returns the resolved node information and cluster group.
```

Comment:
> ... and empty cluster group name

right?

Reply:
I have updated the comment in the most recent commit. 
`If target is a cluster member, it returns the resolved node information and cluster group.` should be the right description.

Reply:
No I dont think that is quite right.

Perhaps the confusion is coming from:

```go
return member, targetGroup, nil
```

Because targetGroup is empty when target is a specific member name.

So really it would be clearer as:

```go
return member, "", nil
```

The group name is not returned when the target is a cluster member.

Reply:
Aaah yes! Now I feel guilty :D Changed.